// next-view-transitions/dist/index.js
'use client';
import { jsx } from 'react/jsx-runtime';
import NextLink from 'next/link';
import { usePathname, useRouter } from 'next/navigation';
import { useRef, useSyncExternalStore, use, useEffect, useState, createContext, useCallback, startTransition } from 'react';

// TODO: This implementation might not be complete when there are nested
// Suspense boundaries during a route transition. But it should work fine for
// the most common use cases.
// This is a global variable to keep track of the view transition state.
let currentViewTransition = null;
function useBrowserNativeTransitions() {
    const pathname = usePathname();
    const currentPathname = useRef(pathname);
    const transition = useSyncExternalStore((callback)=>{
        if (!('startViewTransition' in document)) {
            return ()=>{};
        }
        const onPopState = ()=>{
            let pendingViewTransitionResolve;
            const pendingViewTransition = new Promise((resolve)=>{
                pendingViewTransitionResolve = resolve;
            });
            const pendingStartViewTransition = new Promise((resolve)=>{
                // @ts-ignore
                document.startViewTransition(()=>{
                    resolve();
                    return pendingViewTransition;
                });
            });
            currentViewTransition = [
                pendingStartViewTransition,
                pendingViewTransitionResolve
            ];
            callback();
        };
        window.addEventListener('popstate', onPopState);
        return ()=>{
        // TODO: Intentionally not cleaning up the event listener, otherwise the
        // listener won't be registered again. This might be something related
        // to the `use` call. We should investigate this further.
        };
    }, ()=>currentViewTransition, ()=>null);
    if (transition && currentPathname.current !== pathname) {
        // Whenever the pathname changes, we block the rendering of the new route
        // until the view transition is started (i.e. DOM screenshotted).
        use(transition[0]);
    }
    // Keep the transition reference up-to-date.
    const transitionRef = useRef(transition);
    useEffect(()=>{
        transitionRef.current = transition;
    }, [
        transition
    ]);
    useEffect(()=>{
        // When the new route component is actually mounted, we finish the view
        // transition.
        currentPathname.current = pathname;
        if (transitionRef.current) {
            transitionRef.current[1]();
            transitionRef.current = null;
        }
    }, [
        pathname
    ]);
}

const ViewTransitionsContext = /*#__PURE__*/ createContext(()=>()=>{});
function ViewTransitions({ children }) {
    const [finishViewTransition, setFinishViewTransition] = useState(null);
    useEffect(()=>{
        if (finishViewTransition) {
            finishViewTransition();
            setFinishViewTransition(null);
        }
    }, [
        finishViewTransition
    ]);
    useBrowserNativeTransitions();
    return /*#__PURE__*/ jsx(ViewTransitionsContext.Provider, {
        value: setFinishViewTransition,
        children: children
    });
}
function useSetFinishViewTransition() {
    return use(ViewTransitionsContext);
}

function _extends() {
    _extends = Object.assign || function(target) {
        for(var i = 1; i < arguments.length; i++){
            var source = arguments[i];
            for(var key in source){
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key];
                }
            }
        }
        return target;
    };
    return _extends.apply(this, arguments);
}
// This is a wrapper around next/link that explicitly uses the router APIs
// to navigate, and trigger a view transition.
function Link(props) {
    const router = useRouter();
    const finishViewTransition = useSetFinishViewTransition();
    const onClick = useCallback((e)=>{
        if (props.onClick) {
            props.onClick(e);
        }
        if ('startViewTransition' in document) {
            e.preventDefault();
            // @ts-ignore
            document.startViewTransition(()=>new Promise((resolve)=>{
                    startTransition(()=>{
                        router.push(props.href);
                        finishViewTransition(()=>resolve);
                    });
                }));
        }
    }, [
        props.href,
        props.onClick
    ]);
    return /*#__PURE__*/ jsx(NextLink, _extends({}, props, {
        onClick: onClick
    }));
}

export { Link, ViewTransitions };
